### PR TITLE
Separate errors within http middleware to protected methods

### PR DESCRIPTION
**Pull request by Shane32**

### CODE REVIEW

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs [hunk 1]
```diff
@@ -196,6 +181,57 @@ public async Task InvokeAsync(HttpContext context)
             }
         }
 
+        protected virtual async ValueTask<bool> HandleDeserializationErrorAsync(HttpContext context, Exception ex)
+        {
+            await WriteErrorResponseAsync(
```
Comment (sungam3r) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:186]: `WriteErrorResponseAsync` is private so no one can utilize it as _default path_.

Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs]: Made protected and simplified arguments.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs [hunk 2]
```diff
@@ -62,13 +62,7 @@ public async Task InvokeAsync(HttpContext context)
             if (!isGet && !isPost)
             {
                 httpResponse.Headers["Allow"] = "GET, POST";
-                await WriteErrorResponseAsync(
-                    httpResponse,
-                    serializer,
-                    cancellationToken,
-                    $"Invalid HTTP method. Only GET and POST are supported. {DOCS_URL}",
-                    HttpStatusCode.MethodNotAllowed
```
Comment (sungam3r) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:70]: MethodNotAllowed is lost

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs [hunk 3]
```diff
@@ -114,16 +102,19 @@ public async Task InvokeAsync(HttpContext context)
...
+                        }
                         break;
 
                     default:
-                        await WriteErrorResponseAsync(
-                            httpResponse,
-                            serializer,
-                            cancellationToken,
-                            $"Invalid 'Content-Type' header: non-supported media type. Must be of '{MediaType.JSON}', '{MediaType.GRAPH_QL}' or '{MediaType.FORM}'. {DOCS_URL}",
-                            HttpStatusCode.UnsupportedMediaType);
```
Comment (sungam3r) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:126]: UnsupportedMediaType lost

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs [hunk 4]
```diff
@@ -219,8 +222,7 @@ protected virtual Task RequestExecutedAsync(in GraphQLRequestExecutionResult req
             return Task.CompletedTask;
         }
 
-        private Task WriteErrorResponseAsync(HttpResponse httpResponse, IGraphQLSerializer serializer, CancellationToken cancellationToken,
-            string errorMessage, HttpStatusCode httpStatusCode)
+        protected virtual Task WriteErrorResponseAsync(HttpContext context, string errorMessage, HttpStatusCode httpStatusCode = HttpStatusCode.BadRequest)
```
Comment (sungam3r) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:225]:
```suggestion
        protected virtual Task WriteErrorResponseAsync(HttpContext context, string errorMessage, HttpStatusCode httpStatusCode)
```

That is why in one of the previous PR, I removed the default value.

Comment (sungam3r) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:225]: We have tests but GitHub Actions are broken for now.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs [hunk 5]
```diff
@@ -230,10 +232,10 @@ protected virtual Task RequestExecutedAsync(in GraphQLRequestExecutionResult req
                 }
             };
 
-            httpResponse.ContentType = "application/json";
-            httpResponse.StatusCode = (int)httpStatusCode;
+            context.Response.ContentType = "application/json";
+            context.Response.StatusCode = (int)httpStatusCode;
 
-            return serializer.WriteAsync(httpResponse.Body, result, cancellationToken);
+            return _serializer.WriteAsync(context.Response.Body, result, context.RequestAborted);
```
Comment (sungam3r) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:238]: cancellationToken is lost, see ` var cancellationToken = GetCancellationToken(context);` line

Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs]: That method should not exist.  For now I changed the code to use it.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs (hunk 4 above)
Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:225]: Let's merge the other PR and then you can work on this one.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs (hunk 2 above)
Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:70]: fixed

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs (hunk 3 above)
Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:126]: fixed

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs (hunk 4 above)
Comment (Shane32) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs]: fixed to not be optional

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs [hunk 6]
```diff
@@ -79,12 +73,7 @@ public async Task InvokeAsync(HttpContext context)
             {
                 if (!MediaTypeHeaderValue.TryParse(httpRequest.ContentType, out var mediaTypeHeader))
                 {
-                    await WriteErrorResponseAsync(
-                        httpResponse,
-                        serializer,
-                        cancellationToken,
-                        $"Invalid 'Content-Type' header: value '{httpRequest.ContentType}' could not be parsed.",
-                        HttpStatusCode.BadRequest);
```
Comment (Shane32) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:87]: Changed to media not supported error code.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs (hunk 5 above)
Comment (sungam3r) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs]: Why should not? I use it to link requestaborted token with my custom token to handle user defined operation timeout from request headers.

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs [hunk 7]
```diff
@@ -196,6 +182,24 @@ public async Task InvokeAsync(HttpContext context)
...
+        }
+
+        protected virtual Task HandleNoQueryErrorAsync(HttpContext context)
+            => WriteErrorResponseAsync(context, "GraphQL query is missing.", HttpStatusCode.BadRequest);
+
+        protected virtual Task HandleContentTypeCouldNotBeParsedErrorAsync(HttpContext context)
+            => WriteErrorResponseAsync(context, $"Invalid 'Content-Type' header: value '{context.Request.ContentType}' could not be parsed.", HttpStatusCode.UnsupportedMediaType);
+
+        protected virtual Task HandleInvalidContentTypeErrorAsync(HttpContext context)
+            => WriteErrorResponseAsync(context, $"Invalid 'Content-Type' header: non-supported media type. Must be of '{MediaType.JSON}', '{MediaType.GRAPH_QL}' or '{MediaType.FORM}'. {DOCS_URL}", HttpStatusCode.UnsupportedMediaType);
```
Comment (sungam3r) [inline src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:198]:
```suggestion
            => WriteErrorResponseAsync(context, $"Invalid 'Content-Type' header: non-supported media type '{context.Request.ContentType}'. Must be of '{MediaType.JSON}', '{MediaType.GRAPH_QL}' or '{MediaType.FORM}'. {DOCS_URL}", HttpStatusCode.UnsupportedMediaType);
```

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs (hunk 6 above)
Comment (sungam3r) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs:87]: 👍

File: src/Transports.AspNetCore/GraphQLHttpMiddleware.cs (hunk 5 above)
Comment (sungam3r) [reply src/Transports.AspNetCore/GraphQLHttpMiddleware.cs]: Also see ` protected override CancellationToken GraphQLHttpMiddlewareWithLogs.GetCancellationToken(HttpContext context)`

**Review by sungam3r — APPROVED** [review]
One suggestion regarding error message left.

**Comment by Shane32** [comment]
Waiting for #716

**Review by sungam3r — APPROVED** [review]

**Pull request merged**